### PR TITLE
`fn emu_edge_rust`: Cleanup and make mostly safe

### PR DESCRIPTION
**Pull request by randomPoison**
* Part of #819

### CODE REVIEW

File: src/mc.rs [hunk 1]
```diff
@@ -1004,78 +1006,72 @@ unsafe fn emu_edge_rust<BD: BitDepth>(
...
         );
+        // extend left edge for this line
         if left_ext != 0 {
-            BD::pixel_set(
-                std::slice::from_raw_parts_mut(blk, left_ext as usize),
-                *blk.offset(left_ext as isize),
-                left_ext as usize,
-            );
+            let val = dst[blk + left_ext];
+            dst[blk..][..left_ext].fill(val);
```
Comment (kkysen) [inline src/mc.rs:1045]:
```suggestion
            BD::pixel_set(&dst[blk..], val, left_ext);
```

I know these do the same thing, but maybe we should stick with the naming C uses?  That said, this way is indeed easier to read, so your call.

File: src/mc.rs [hunk 2]
```diff
@@ -1004,78 +1006,72 @@ unsafe fn emu_edge_rust<BD: BitDepth>(
...
-            BD::pixel_set(
-                std::slice::from_raw_parts_mut(
-                    blk.offset(left_ext as isize).offset(center_w as isize),
-                    right_ext as usize,
-                ),
-                *blk.offset((left_ext + center_w - 1) as isize),
-                right_ext as usize,
-            );
+            let val = dst[blk + left_ext + center_w - 1];
+            dst[blk + left_ext + center_w..][..right_ext].fill(val);
```
Comment (kkysen) [inline src/mc.rs:1050]:
```suggestion
            BD::pixel_set(&dst[blk + left_ext + center_w..], val, right_ext);
```

File: src/mc.rs [hunk 3]
```diff
@@ -1004,78 +1006,72 @@ unsafe fn emu_edge_rust<BD: BitDepth>(
...
-            bw as usize,
-        );
-        dst = dst.offset(BD::pxstride(dst_stride));
-        y_1 += 1;
+
+    // copy top
+    let mut dst_off = 0;
+    let blk = top_ext * dst_stride;
+    for _ in 0..top_ext {
+        let (front, back) = dst.split_at_mut(blk);
```
Comment (kkysen) [inline src/mc.rs:1060]: This can be outside of the loop, right?

File: src/mc.rs [hunk 4]
```diff
@@ -1004,78 +1006,72 @@ unsafe fn emu_edge_rust<BD: BitDepth>(
...
+        BD::pixel_copy(&mut front[dst_off..][..bw], &mut back[..bw], bw);
```
Comment (kkysen) [inline src/mc.rs:1061]:
```suggestion
        BD::pixel_copy(&mut front[dst_off..][..bw], &back[..bw], bw);
```

File: src/mc.rs [hunk 5]
```diff
@@ -1004,78 +1006,72 @@ unsafe fn emu_edge_rust<BD: BitDepth>(
...
+    // copy bottom
+    dst_off += center_h * dst_stride;
+    for _ in 0..bottom_ext {
+        let (front, back) = dst.split_at_mut(dst_off);
         BD::pixel_copy(
-            std::slice::from_raw_parts_mut(dst, bw as usize),
-            std::slice::from_raw_parts(dst.offset(-BD::pxstride(dst_stride)), bw as usize),
-            bw as usize,
+            &mut back[..bw],
+            &mut front[dst_off - dst_stride..][..bw],
```
Comment (kkysen) [inline src/mc.rs:1071]:
```suggestion
            &front[dst_off - dst_stride..][..bw],
```

**Review by kkysen — CHANGES_REQUESTED** [review]
Couple of very little things; rest all LGTM.

File: src/mc.rs (hunk 1 above)
Comment (randomPoison) [reply src/mc.rs]: Eh yeah I can go back to using `pixel_set`, it doesn't matter much either way.

File: src/mc.rs (hunk 3 above)
Comment (randomPoison) [reply src/mc.rs]: Yup, moved.

**Review by kkysen — APPROVED** [review]

**Pull request merged**